presto-main/src/main/java/io/prestosql/operator/aggregation/ApproximateMostFrequentHistogram.java

```diff
@@ -0,0 +1,139 @@
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(ApproximateMostFrequentHistogram.class).instanceSize();
+    // Larger capacity for stream summary improves the accuracy.
+    private static final int MAX_CAPACITY_FACTOR = 5;
+
+    private StreamSummary<K> streamSummary;
+    private int maxBuckets;
+    private int capacity;
+    private ApproximateMostFrequentBucketSerializer<K> serializer;
+    private ApproximateMostFrequentBucketDeserializer<K> deserializer;
+
+    /**
+     * @param maxBuckets The maximum number of elements stored in the bucket.
+     * @param capacity The maximum capacity of the stream summary data structure.
+     * @param serializer It serializes a bucket into varbinary slice.
+     * @param deserializer It appends a bucket into the histogram.
+     */
+    public ApproximateMostFrequentHistogram(int maxBuckets,
+                                            int capacity,
+                                            ApproximateMostFrequentBucketSerializer<K> serializer,
+                                            ApproximateMostFrequentBucketDeserializer<K> deserializer)
+    {
+        requireNonNull(serializer, "serializer is null");
+        requireNonNull(deserializer, "deserializer is null");
+        streamSummary = new StreamSummary<>(capacity);
+        this.maxBuckets = maxBuckets;
+        this.capacity = capacity;
+        this.serializer = serializer;
+        this.deserializer = deserializer;
+    }
+
+    public ApproximateMostFrequentHistogram(Slice serialized,
+                                            ApproximateMostFrequentBucketSerializer<K> serializer,
+                                            ApproximateMostFrequentBucketDeserializer<K> deserializer)
+    {
+        SliceInput input = serialized.getInput();
+
+        checkArgument(input.readByte() == FORMAT_TAG, "Unsupported format tag");
+
+        this.maxBuckets = input.readInt();
+        this.capacity = input.readInt();
+        this.streamSummary = new StreamSummary<>(capacity);
+        this.serializer = serializer;
+        this.deserializer = deserializer;
+
+        for (int i = 0; i < maxBuckets; i++) {
+            this.deserializer.deserialize(input, this);
+        }
+    }
+
+    public void add(K value)
+    {
+        streamSummary.offer(value);
+    }
+
+    public void add(K value, long incrementCount)
+    {
+        streamSummary.offer(value, toIntExact(incrementCount));
+    }
+
+    public Slice serialize()
+    {
+        DynamicSliceOutput output = new DynamicSliceOutput(1000);
+        List<Counter<K>> counters = streamSummary.topK(maxBuckets);
+        output.appendByte(FORMAT_TAG);
+        output.appendInt(maxBuckets);
+        output.appendInt(capacity);
+        // Serialize key and counts.
+        for (Counter<K> counter : counters) {
+            serializer.serialize(counter.getItem(), counter.getCount(), output);
+        }
+
+        return output.getUnderlyingSlice();
```

Comment:
This will return the raw allocated slice, which may be bigger than needed. Use `output.slice()` instead.